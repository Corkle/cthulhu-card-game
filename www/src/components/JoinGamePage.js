import React, { useState } from 'react'
import PropTypes from 'prop-types'
import * as API from '../api'


/*

const UserLogin = (props) => {
	const [loginState, setLoginState] = useState('INITIAL')
  const [usernameText, setUsernameText] = useState('')
  const [usernameTextError, setUsernameTextError] = useState(null)
  const [gameIdText, setGameIdText] = useState('')
  const [gameIdTextError, setGameIdTextError] = useState(null)
  const [gameId, setGameId] = useState('')

  const handleUsernameTextChange = (e) => {
    setUsernameText(e.target.value)
    setUsernameTextError(null)
  }

  const handleGameIdTextChange = (e) => {
    setGameIdText(e.target.value)
    setGameIdTextError(null)
  }

  const handleNewGameClick = (e) => {
    if (usernameText != '') {
      API.createNewGame(usernameText, (data, err) => {
        if (!err) {
          props.onGameJoined(data)
          setGameId(data.game.gameId)
          setLoginState('NEW_GAME_SUCCESS')
        }
      })
    }
  }

  const handleJoinGameClick = (e) => {
    if (gameIdText != '') {
      API.joinGameById(gameIdText, usernameText, (data, err) => {
        if (!err) {
          props.onGameJoined(data)
          props.onStartPlaying()
        } else {
          setGameIdTextError(err)
        }
      })
    } else {
      setGameIdTextError('Enter a valid Game ID') 
    }
  }

  const handleStartClick = (e) => {
    usernameText != '' ?
      setLoginState('CONNECT_TYPE') :
      setUsernameTextError('Enter a valid username')
  }

	const LOGIN_STATES = {
		'LOADING':
			<div className='has-text-centered'>
				Loading...
			</div>
		,
		'INITIAL':
      <> 
        <BigTextbox
          placeholder={'Username'}
          error={usernameTextError}
          onChange={handleUsernameTextChange}
          onPressEnter={handleStartClick} />
        <BigButton text={'Start'} onClick={handleStartClick} />
      </>,
		'CONNECT_TYPE':
      <>
        <BigButton text={'New Game'} onClick={handleNewGameClick} />
        <BigButton text={'Join Game'} onClick={ (e) => setLoginState('JOIN_GAME') } />
      </>,
    'JOIN_GAME':
      <>
        <BigTextbox
          placeholder={'Game ID'}
          error={gameIdTextError}
          onChange={handleGameIdTextChange}
          onPressEnter={handleJoinGameClick} />
        <BigButton text={'Join'} onClick={handleJoinGameClick}/>
      </>,
    'NEW_GAME_SUCCESS':
      <>
        <div className='box has-text-centered'>
          <p>Friends can join using this Game ID:</p>
          <h1 className='content is-large'>{ gameId }</h1>
        </div>
        <BigButton text={'OK'} onClick={props.onStartPlaying} />
      </>
	}

  return (
    <div className='section is-large'>
      <div className='column is-4 is-offset-4'>
				<div className='box'>
          { LOGIN_STATES[loginState] }
        </div>
      </div>
    </div>
  )
}
*/

const BigInputText = (props) => {
  return (
    <div className='field'>
      <div className='control'>
        <input
          className='input is-large'
          type='text'
          placeholder={ props.placeholder }
          onChange={ props.onChange }
          onKeyDown={ (e) => { if (e.keyCode == 13) { props.onPressEnter() }} }
        />
      </div>
      {props.error &&
        <p className='help is-danger'>{props.error}</p>
      } 
    </div>
  )
}

BigInputText.propTypes = {
  error: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onPressEnter: PropTypes.func.isRequired,
  placeholder: PropTypes.string.isRequired
}

const BigButton = (props) => {
  return (
    <div className='field'>
      <button
        className={ 'button is-info is-large is-outlined is-fullwidth' }
        onClick={ props.onClick }>
        { props.text }
      </button>
    </div>
  )
}

BigButton.propTypes = {
  onClick: PropTypes.func,
  text: PropTypes.string
}

const JoinGamePage = (props) => {
  const [nicknameText, setNicknameText] = useState('')
  const [nicknameTextError, setNicknameTextError] = useState(null)

  const handleNextClick = (e) => {
    if (nicknameText == '') {
      setNicknameTextError('Please enter a valid nickname')
    }
  }

  return (
    <div className='section is-large'>
      <div className='column is-4 is-offset-4'>
				<div className='box has-text-centered'>
          <h3 className='subtitle is-4'>Choose a nickname</h3>
          <BigInputText
            placeholder='Nickname'
            onChange={(e) => {console.log(e)}}
            onPressEnter={() => {}}
            error={nicknameTextError} />
          <BigButton text='Next' onClick={handleNextClick} />
        </div>
      </div>
    </div>
  )
}

JoinGamePage.propTypes = {
  //onGameJoined: PropTypes.func.isRequired,
  //onStartPlaying: PropTypes.func.isRequired
}

export default JoinGamePage
