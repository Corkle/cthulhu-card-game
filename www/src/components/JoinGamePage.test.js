import React from 'react'
import { create } from 'react-test-renderer'
import JoinGamePage from './JoinGamePage'
jest.mock('../socket')

const createComponent = () => {
  return create(<JoinGamePage />)
}

describe('Nickname selection view', () => {
  describe('initial load', () => {
    test('should render component', () => {
      const component = createComponent() 
      expect(component).toMatchSnapshot()
    })

    test('should render nickname text input', () => {
      const component = createComponent() 
      const input = component.root.findByType('input')
      expect(input.props.placeholder).toEqual('Nickname')
      expect(input.props.type).toEqual('text')
      expect(input.props.className).toEqual('input is-large')
      expect(input.props.onChange).toBeDefined()
      expect(input.props.onKeyDown).toBeDefined()
    })

    test('should render Next button', () => {
      const component = createComponent() 
      const button = component.root.findByType('button')
      expect(button.props.children).toEqual('Next')
      expect(button.props.onClick).toBeDefined()
    })
  })

  describe('clicking Next button', () => {
    test('when nickname is blank, should show error text', () => {
      const component = createComponent()
      const input = component.root.findByType('input')
      const button = component.root.findByType('button')
      input.props.onChange({target: {value: ''}})
      button.props.onClick()
      const errorText = component.root.findByType('p')
      expect(errorText.props.children).toEqual('Please enter a valid nickname')
    })
  })
})
