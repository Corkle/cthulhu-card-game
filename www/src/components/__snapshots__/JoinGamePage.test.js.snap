// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nickname selection view initial load should render component 1`] = `
<div
  className="section is-large"
>
  <div
    className="column is-4 is-offset-4"
  >
    <div
      className="box has-text-centered"
    >
      <h3
        className="subtitle is-4"
      >
        Choose a nickname
      </h3>
      <div
        className="field"
      >
        <div
          className="control"
        >
          <input
            className="input is-large"
            onChange={[Function]}
            onKeyDown={[Function]}
            placeholder="Nickname"
            type="text"
          />
        </div>
      </div>
      <div
        className="field"
      >
        <button
          className="button is-info is-large is-outlined is-fullwidth"
          onClick={[Function]}
        >
          Next
        </button>
      </div>
    </div>
  </div>
</div>
`;
